### PR TITLE
refactor(kclvm-runner): encapsulate dylib generating, linking and executing in kclvm/lib.rs into kclvm-runner

### LINKED ISSUE #67: KCL Validation tool (Rust ver.)
## Enhancement

According to the content in the Roadmap https://github.com/KusionStack/KCLVM/issues/29 , use Rust to rewrite the KCL validation tool.

### PR DESCRIPTION
**Pull request by zong-zhe**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

fix #67 and #82

#### 2. What is the scope of this PR (e.g. component or file name):
kclvm-runner/runner.rs
kclvm-runner/lib.rs

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

1. Encapsulate dylibs generating and executing in kclvm/lib.rs into kclvm-runner.
2. Add struct "KclvmAssembler" in kclvm-runner/runner.rs to provide method "gen_libs" for dynamic link library generating.
3. Add struct "KclvmLinker" in kclvm-runner/runner.rs to provide method "link_all_dylibs" for dynamic link library linking.
4. Add method "execute" in kclvm-runner/lib.rs to encapsulate dylibs generating(gen_libs), dynamic link library linking(link_all_libs) and running(runner.run) together.
5. The main purpose of separating the three parts from kclvm/lib.rs and encapsulating them into kclvm-runner is to support reuse in kcl-vet.
6. Decoupling the assembler and llvm.
7. The assembling LLVM IR into a dynamic link library is encapsulated into "LlvmLibAssembler" separately.
8. Add trait "LibAssembler" to describe the interface "KclvmLibAssembler" should have. If other intermediate code is added in the future, the corresponding assembler must implement this trait.
9. Struct "LlvmLibAssembler" is provided in "KclvmLibAssembler".  "KclvmLibAssembler" is an enum that implements trait "LibAssembler". "KclvmLibAssembler" is responsible for the compilation of a single kcl file in one thread.
10. "KclvmAssembler" is responsible for the concurrent compilation of multiple kcl files.
11. "KclvmAssembler" will call the method in "KclvmLibAssembler" to generate a dynamic link library for a single kcl file in each thread of concurrent compilation.

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

Add test cases in kclvm-runner/src/tests.rs and kclvm-runner/test_datas.
Add bench in kclvm-runner/benches/bench_runner.rs

###### 1. KCL Test Case
```
// kclvm/runner/src/test_datas/init_check_order_0/main.k
schema Person:
    name: str
    age: int
    gender: str
    info: str = "{}, {}, {} years old".format(name, gender, age)

alice = Person {
    "name": "alice",
    "age": 10,
    "gender": "female"
}
```

###### 2.1 CLI Report 
image

###### 2.2 HTML Report
image
image

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: kclvm/runner/src/runner.rs [hunk 1]
```diff
@@ -79,6 +92,158 @@ pub struct KclvmRunnerOptions {
     pub plugin_agent_ptr: u64,
 }
 
+/// KclvmAssembler is mainly responsible for generating bytecode,
+/// LLVM IR or dylib, and take the result of kclvm-parser and kclvm-sema
+/// as input.
+pub struct KclvmAssembler {}
```
Comment (Peefy) [inline kclvm/runner/src/runner.rs:98]: Move `KclvmAssembler` and `KclvmLinker` to another mod?

File: kclvm/runner/src/lib.rs [hunk 2]
```diff
@@ -1,2 +1,42 @@
...
+/// scope: ProgramScope returned by kclvm-sema
+///
+/// In the method, dylibs is generated by KclvmAssembler, and method "KclvmAssembler::gen_dylibs"
+/// will return dylibs path in a "Vec<String>";
+///
+/// After linking all dylibs by KclvmLinker, method "KclvmLinker::link_all_dylibs" will return a path
+/// for dylib.
+///
+/// At last, KclvmRunner will be constructed and call method "run" to execute the kcl program.
+pub fn execute(
```
Comment (Peefy) [inline kclvm/runner/src/lib.rs:22]: 1. Add some Samples directly in the doc comment for testing
2. This end-to-end function can add some benchmarks

File: kclvm/runner/src/assembler.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+use kclvm_ast::ast;
+use kclvm_compiler::codegen::{llvm::emit_code, EmitOptions};
+use kclvm_config::cache::{load_pkg_cache, save_pkg_cache, CacheOption};
+use kclvm_sema::resolver::scope::ProgramScope;
+
+/// KclvmAssembler is mainly responsible for generating bytecode,
+/// LLVM IR or dylib, and take the result of kclvm-parser and kclvm-sema
+/// as input.
+pub struct KclvmAssembler {}
+const LL_FILE: &str = "_a.out";
```
Comment (Peefy) [inline kclvm/runner/src/assembler.rs:19]: move `LL_FILE` to top level const definitions after `use` statements.

File: kclvm/runner/src/lib.rs [hunk 4]
```diff
@@ -1,2 +1,53 @@
...
+/// "args" is the items selected by the user in the KCLVM CLI.
+///
+/// In the method, dylibs is generated by KclvmAssembler, and method "KclvmAssembler::gen_dylibs"
+/// will return dylibs path in a "Vec<String>";
+///
+/// After linking all dylibs by KclvmLinker, method "KclvmLinker::link_all_dylibs" will return a path
+/// for dylib.
+///
+/// At last, KclvmRunner will be constructed and call method "run" to execute the kcl program.
+pub fn execute(
```
Comment (Peefy) [inline kclvm/runner/src/lib.rs:33]: Add some Samples directly in the doc comment for testing using the ``` code block. For example:
```rust
use kclvm::runner::execute;
// Omit other usages

fn main() {
    let result = execute(..).unwrap();
    assert_eq!(result, ...);
}
```

**Comment by ldxdl** [comment]
Pls put benchmark stats into the PR info.

File: kclvm/runner/src/assembler.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                pkgs,
+                cmd_args: vec![],
+                cmd_overrides: vec![],
+            };
+            compile_progs.insert(
+                pkgpath,
+                (compile_prog, scope.import_names.clone(), cache_dir.clone()),
+            );
+        }
+        let pool = ThreadPool::new(4);
```
Comment (ldxdl) [inline kclvm/runner/src/assembler.rs:80]: Why is here a hardcoded thread count?

File: kclvm/runner/src/assembler.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+use kclvm_ast::ast;
+use kclvm_compiler::codegen::{llvm::emit_code, EmitOptions};
+use kclvm_config::cache::{load_pkg_cache, save_pkg_cache, CacheOption};
+use kclvm_sema::resolver::scope::ProgramScope;
+
+/// LLVM IR file suffix.
+const LL_FILE: &str = "_a.out";
+
+/// KclvmAssembler is mainly responsible for generating bytecode,
+/// LLVM IR or dylib, and take the result of kclvm-parser and kclvm-sema
```
Comment (ldxdl) [inline kclvm/runner/src/assembler.rs:19]: Ultimately the assembler should not be directly aware of llvm.

File: kclvm/runner/src/assembler.rs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+                            from_path: None,
+                            emit_path: Some(&ll_file),
+                            no_link: true,
+                        },
+                    )
+                    .expect("Compile KCL to LLVM error");
+                    let mut cmd = Command::new(plugin_agent);
+                    cmd.run_clang_single(&ll_path, &dylib_path)
+                } else {
+                    // If AST module has been modified, ignore the dylib cache
```
Comment (ldxdl) [inline kclvm/runner/src/assembler.rs:116]: Why ATS has beed modified here?

File: kclvm/runner/src/assembler.rs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+                    match dylib_relative_path {
+                        Some(dylib_relative_path) => {
+                            if dylib_relative_path.starts_with('.') {
+                                dylib_relative_path.replacen(".", root, 1)
+                            } else {
+                                dylib_relative_path
+                            }
+                        }
+                        None => {
+                            emit_code(
```
Comment (ldxdl) [inline kclvm/runner/src/assembler.rs:128]: reuse emit_code part of code?

**Review by ldxdl — COMMENTED** [review]
As a refactor PR, it would be good to have some preparation for future backend target abstractions.

File: kclvm/runner/src/assembler.rs (hunk 5 above)
Comment (zong-zhe) [reply kclvm/runner/src/assembler.rs]: I changed the thread count to a member of structure KclvmAssembler, and added the constructor "new" and "new_with_thread_count" to assign value ​​to this member, its default value is 4.

File: kclvm/runner/src/assembler.rs (hunk 7 above)
Comment (zong-zhe) [reply kclvm/runner/src/assembler.rs]: This comment is left from the previous code and has been corrected.

File: kclvm/runner/src/assembler.rs (hunk 6 above)
Comment (zong-zhe) [reply kclvm/runner/src/assembler.rs]: I modified the comments for KclvmAssembler and encapsulated the part that generates the llvm code in static method "lock_file_and_gen_dylib".

File: kclvm/runner/src/assembler.rs (hunk 5 above)
Comment (ldxdl) [reply kclvm/runner/src/assembler.rs]: Could you give it a default value based on the number of cores on the machine?

File: kclvm/runner/src/assembler.rs [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+            };
+            compile_progs.insert(
+                pkgpath,
+                (compile_prog, scope.import_names.clone(), cache_dir.clone()),
+            );
+        }
+        let pool = ThreadPool::new(self.thread_count);
+        let (tx, rx) = channel();
+        let prog_count = compile_progs.len();
+        // let temp_entry_file = temp_file();
```
Comment (ldxdl) [inline kclvm/runner/src/assembler.rs:213]: remove unused code

File: kclvm/runner/src/assembler.rs [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+use threadpool::ThreadPool;
+
+use crate::command::Command;
+use kclvm_ast::ast::{self, Program};
+use kclvm_compiler::codegen::{llvm::emit_code, EmitOptions};
+use kclvm_config::cache::{load_pkg_cache, save_pkg_cache, CacheOption};
+use kclvm_sema::resolver::scope::ProgramScope;
+
+/// LLVM IR file suffix.
+const LL_FILE: &str = "_a.out";
```
Comment (ldxdl) [inline kclvm/runner/src/assembler.rs:17]: I prefer rename it to IR_FILE, and make it a default IR file of KclvmAssembler

**Comment by ldxdl** [comment]
This refactoring should preferably do some preparation for the subsequent target abstraction refactoring. For example, dylib and llvm were all hardcoded as the only impl way, you may need to change them into variables first, and then the next refactoring will replace these variables through traits or other dynamic methods to make it a target abstraction with a set of impls. At the same time, do not hardcode implementation keywords like  'dylib', 'llvm' in the comments and docstrings.

**Review by Peefy — COMMENTED** [review]
You'd better lint code by using `cargo clippy` in the `runner` crate.

**Comment by ldxdl** [comment]
Globally search for dylibs, LLVM, and see if it makes sense.

File: internal/kclvm_py/scripts/update-kclvm.sh [hunk 11]
```diff
@@ -122,6 +122,7 @@ cargo build --release --target wasm32-unknown-unknown
 cp $topdir/kclvm/target/wasm32-unknown-unknown/release/libkclvm.a $kclvm_install_dir/lib/libkclvm_wasm32.a
 cp src/_kclvm_undefined_wasm.txt $kclvm_install_dir/lib/_kclvm_undefined_wasm.txt
 
+
```
Comment (Peefy) [inline internal/kclvm_py/scripts/update-kclvm.sh:125]: revert the diff

File: kclvm/runner/src/lib.rs [hunk 12]
```diff
@@ -1,2 +1,133 @@
...
+
+#[inline]
+pub fn remove_file(file: &str) {
+    if Path::new(&file).exists() {
+        std::fs::remove_file(&file).unwrap();
+    }
+}
+
+/// Returns a temporary file name consisting of timestamp and process id.
+pub fn temp_file(dir: &str) -> String {
```
Comment (Peefy) [inline kclvm/runner/src/lib.rs:127]: use private function `pub fn` -> `fn`

File: kclvm/runner/src/lib.rs [hunk 13]
```diff
@@ -1,2 +1,133 @@
...
+
+/// Clean all the tmp files generated during lib generating and linking.
+#[inline]
+pub fn clean_tmp_files(temp_entry_file: &String, lib_suffix: &String) {
+    let temp_entry_lib_file = format!("{}{}", temp_entry_file, lib_suffix);
+    remove_file(&temp_entry_lib_file);
+}
+
+#[inline]
+pub fn remove_file(file: &str) {
```
Comment (Peefy) [inline kclvm/runner/src/lib.rs:120]: use private function `pub fn -> fn`

File: kclvm/runner/src/lib.rs [hunk 14]
```diff
@@ -1,2 +1,133 @@
...
+
+    // Clean temp files
+    remove_file(&lib_path);
+    clean_tmp_files(&temp_entry_file, &lib_suffix);
+    result
+}
+
+/// Clean all the tmp files generated during lib generating and linking.
+#[inline]
+pub fn clean_tmp_files(temp_entry_file: &String, lib_suffix: &String) {
```
Comment (Peefy) [inline kclvm/runner/src/lib.rs:114]: use private function `pub fn -> fn`

File: kclvm/runner/Cargo.toml [hunk 15]
```diff
@@ -23,3 +25,17 @@ kclvm-config = {path = "../config", version = "0.1.0"}
...
 kclvm-sema = {path = "../sema", version = "0.1.0"}
 kclvm-version = {path = "../version", version = "0.1.0"}
+kclvm-error = {path = "../error", version="0.1.0"}
+kclvm-tools = {path = "../tools", version = "0.1.0"}
+
+[dev-dependencies]
+kclvm-parser = {path = "../parser", version = "0.1.0"}
+criterion = "0.3"
+
+# [[bench]]
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply kclvm/runner/Cargo.toml]: Deleted

File: kclvm/runner/src/lib.rs (hunk 12 above)
Comment (zong-zhe) [reply kclvm/runner/src/lib.rs]: I called this function in unit test, so I use `pub fn` -> `pub(crate) fn`.

File: kclvm/runner/src/lib.rs (hunk 13 above)
Comment (zong-zhe) [reply kclvm/runner/src/lib.rs]: I called this function in unit test, so I use pub fn -> pub(crate) fn.

File: kclvm/runner/src/lib.rs (hunk 14 above)
Comment (zong-zhe) [reply kclvm/runner/src/lib.rs]: I called this function in unit test, so I use pub fn -> pub(crate) fn.

**Review by Peefy — DISMISSED** [review]
LGTM

**Review by chai2010 — APPROVED** [review]
LGTM

**Pull request merged**